cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowPrivate.qll

```diff
@@ -412,16 +365,73 @@ private predicate exactReadStep(Node node1, ArrayContent a, Node node2) {
   )
 }
 
+pragma[noinline]
+private predicate getField(FieldContent fc, Field field) {
+  exists(Class c |
+    c = field.getDeclaringType() and
+    fc.getAField() = field and
+    fc.hasOffset(c, _, _)
+  )
+}
+
+/**
+ * Given two nodes `node1` and `node2`, the shared dataflow library infers a store step
+ * from `node1` to `node2` is there is a read step from the the pre update node of `node2` to the
+ * the pre update node of `node1`. This ensures that a field assignment such as
+ * ```cpp
+ * x.y.z = source();
+ * ```
+ * generates the access path `[z, y, x]`. This predicate ensures that there's a read step from
+ * `x` (the pre update node of `y`) to `y` (the pre update node of `z`), which means that the
+ * shared dataflow library will infer a store step from `z` to `y`.
```

Comment:
In AST flow it's been convenient that the shared data-flow library adds "reverse read steps" (store steps) automatically because those reverse steps are derived from forward steps that have value on their own and would be there in any case. If it's a hassle to add the forward steps in IR data flow, and if we don't actually need them in their forward form, could we then just extend `storeStep` ourselves instead of relying on reverse reads?

Reply:
We could do that, yes. I don't think one is more of a hassle than the other. As it's currently implemented in this PR, we still need the `reverseReadStep` predicate to pop `c` off the access path in `sink(a.b.c)` (here, the pre-update node of `c` is `b`, and so `reverseReadStep(b, _, c)` holds)